internal/vulnstore/postgres/gc.go

```diff
@@ -0,0 +1,91 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+const (
+	// GCThrottle sets a limit for the number of update
+	// operations deleted in a single GC run.
+	GCThrottle = 3
+)
+
+func gc(ctx context.Context, pool *pgxpool.Pool, keep int) (int64, int64, error) {
+	const (
+		// this query will return rows of string,[]uuid.UUID where string is the updater's name
+		// and []uuid.UUID is a slice of update operation refs exceeding the provided keep value.
+		updateOps = `
+WITH ordered_ops AS (
+    SELECT updater, array_agg(ref ORDER BY date DESC) AS refs FROM update_operation GROUP BY updater
+)
+SELECT ordered_ops.updater, ordered_ops.refs[$1:]
+FROM ordered_ops
+WHERE array_length(ordered_ops.refs, 1) > $2;
+`
+
+		deleteVulns = `
+DELETE FROM vuln WHERE id NOT IN (SELECT vuln FROM uo_vuln);
+`
+	)
+
+	// keys are updater names, values are a slice of stale
+	// update operations.
+	m := map[string][]uuid.UUID{}
+
+	// gather any update operations exceeding our keep value.
+	// keep+1 is used because PG's array slicing is inclusive,
+	// we want to grab all items once after our keep value.
+	rows, err := pool.Query(ctx, updateOps, keep+1, keep)
+	switch err {
+	case nil:
+	case pgx.ErrNoRows:
+		return 0, 0, nil
+	default:
+		return 0, 0, fmt.Errorf("error querying for update operations: %w", err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var updater string
+		// pgx will not scan directly into a []uuid.UUID
+		tmp := pgtype.UUIDArray{}
+		err := rows.Scan(&updater, &tmp)
+		if err != nil {
+			return 0, 0, fmt.Errorf("error scanning update operations: %w", err)
+		}
+		for _, u := range tmp.Elements {
+			m[updater] = append(m[updater], u.Bytes) // this works since [16]byte value is assignable to uuid.UUID
+		}
+
+	}
+
+	// delete update operations up until GCThrottle
+	var i int
+	for updater, refs := range m {
+		if i >= GCThrottle {
+			break
+		}
+		err := deleteUpdateOperations(ctx, pool, refs...)
+		if err != nil {
+			return 0, 0, fmt.Errorf("error deleting update operations: %w", err)
+		}
+		delete(m, updater)
+		i++
```

Comment:
What happens when two updaters race on issuing this delete? Is that actually an error? Should it be counted against the number of deletes allowed?

Reply:
This delete is done in a single query which means a single implicit transaction I think. One will block on the other, and the later's query will result in no rows to delete. Correct me if I'm wrong.

Reply:
Right, but that means the delete will error when the locks are released. Should that end the entire GC call, as it does here?

Reply:
Wont the second transaction *not even start* until the first one is finished? If that's the case, the second one will just find nothing to delete, not throw an error.